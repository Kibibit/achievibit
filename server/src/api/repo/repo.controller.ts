import {
  Controller,
  Logger,
  NotFoundException,
  Param,
  UseFilters
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';

import { GetAll, GetOne } from '@kb-decorators';
import { KbValidationExceptionFilter } from '@kb-filters';
import { Repo } from '@kb-models';

import { RepoService } from './repo.service';

@Controller('api/repo')
@ApiTags('repo')
@UseFilters(new KbValidationExceptionFilter())
export class RepoController {
  private readonly logger = new Logger(RepoController.name);

  constructor(private readonly repoService: RepoService) {}

  @GetAll(Repo)
  async getAllRepos() {
    const repos = await this.repoService.findAllRepos();

    return repos;
  }

  @GetOne(Repo, ':name')
  async getRepo(@Param('name') name: string) {
    const repo = await this.repoService.findByName(name);

    if (!repo) {
      throw new NotFoundException(`Repo with name ${ name } not found`);
    }

    // will show secret fields as well!
    this.logger.log('Full Repo');
    // will log only public fields!
    this.logger.log(repo);
    // DANGER! WILL LOG EVERYTHING!
    // console.log(repo);

    // will only include exposed fields
    return repo;
  }
}
