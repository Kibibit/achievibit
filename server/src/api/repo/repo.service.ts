import { ReturnModelType } from '@typegoose/typegoose';

import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';

import { BaseService } from '@kb-abstracts';
import { Repo } from '@kb-models';

@Injectable()
export class RepoService extends BaseService<Repo> {
  constructor(
    @InjectModel(Repo.modelName)
    private readonly repoModel: ReturnModelType<typeof Repo>
  ) {
    super(repoModel, Repo);
  }

  async findAllRepos(): Promise<Repo[]> {
    const dbRepos = await this.findAll().exec();

    return dbRepos.map((repo) => new Repo(repo.toObject()));
  }

  async findByName(name: string): Promise<Repo> {
    const dbRepo = await this.findOne({ name }).exec();

    if (!dbRepo) {
      return;
    }

    return new Repo(dbRepo.toObject());
  }
}
