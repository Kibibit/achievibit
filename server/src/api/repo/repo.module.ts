import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';

import { Repo } from '@kb-models';

import { RepoController } from './repo.controller';
import { RepoService } from './repo.service';

@Module({
  imports: [
    MongooseModule.forFeature([
      { name: Repo.modelName, schema: Repo.schema }
    ])
  ],
  providers: [RepoService],
  controllers: [RepoController],
  exports: [RepoService]
})
export class RepoModule {}
