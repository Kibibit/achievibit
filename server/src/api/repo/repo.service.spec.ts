import { MongooseModule } from '@nestjs/mongoose';
import { Test, TestingModule } from '@nestjs/testing';

import {
  closeInMemoryDatabaseConnection,
  createInMemoryDatabaseModule,
  DtoMockGenerator
} from '@kb-dev-tools';
import { Repo } from '@kb-models';

import { RepoService } from './repo.service';

describe('RepoService', () => {
  let service: RepoService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        createInMemoryDatabaseModule(),
        MongooseModule.forFeature([{
          name: Repo.modelName,
          schema: Repo.schema
        }])
      ],
      providers: [RepoService]
    }).compile();

    service = module.get<RepoService>(RepoService);
  });

  afterEach(async () => {
    await closeInMemoryDatabaseConnection();
  });

  it('should be defined', () => {
    expect(service).toBeDefined(); 
  });

  it('should be able to get all repos', async () => {
    const repo1 = DtoMockGenerator.repo();
    const repo2 = DtoMockGenerator.repo();

    const createdRepo1 = await service.create(repo1);
    const createdRepo2 = await service.create(repo2);

    const foundRepos = await service.findAllRepos();

    expect(foundRepos).toContainEqual(createdRepo1);
    expect(foundRepos).toContainEqual(createdRepo2);
    expect(foundRepos.length).toBeGreaterThanOrEqual(2);
  });

  it('should be able to get a repo by name', async () => {
    const repo = DtoMockGenerator.repo();

    const createdRepo = await service.create(repo);

    const foundRepo = await service.findByName(repo.name);

    expect(foundRepo).toEqual(createdRepo);
  }); 

  it('should return null if repo does not exist', async () => {
    const foundRepo = await service.findByName('name');

    expect(foundRepo).toBeUndefined();
  });

  it('should NOT allow creating 2 repos with same fullname', async () => {
    const repo1 = new Repo({
      ...DtoMockGenerator.repo(),
      fullname: 'test-user/test-repo'
    });
    const repo2 = new Repo({
      ...DtoMockGenerator.repo(),
      fullname: 'test-user/test-repo'
    });

    const createdRepo = await service.create(repo1);
    let creationError: string;
    await service.create(repo2)
      .catch((err) => creationError = err.message);

    expect(creationError).toMatch(/E11000 duplicate key error/);
    expect(creationError).toMatch(/test-user\/test-repo/);

    const allRepos = await service.findAllRepos();
    expect(allRepos.length).toBe(1);
    expect(allRepos[0]).toEqual(expect.objectContaining(createdRepo));
    expect(allRepos[0]._id).toBeDefined()
    expect(allRepos[0]._id).toEqual(createdRepo._id);
  });
});
