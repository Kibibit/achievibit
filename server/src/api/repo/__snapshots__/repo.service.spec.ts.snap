// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoService should NOT allow creating 2 repos with same fullname 1`] = `"E11000 duplicate key error collection: repos index: fullname_1 dup key: { fullname: \\"test-user/test-repo\\" }"`;
